torch/_subclasses/fake_tensor.py

```diff
@@ -905,13 +905,16 @@ def device(self):
 
     @staticmethod
     def __new__(cls, fake_mode, elem, device, constant=None):
-        self = torch.Tensor._make_subclass(
-            cls,
-            elem,
-            elem.requires_grad,
-            dispatch_device=True,
-            device_for_backend_keys=device,
-        )
+        if isinstance(elem, torch._subclasses.FakeTensor):
```

Comment:
nit: don't recall, is this still needed?

Reply:
yup, otherwise it raises `RuntimeError: Creating a new Tensor subclass FakeTensor but the raw Tensor object is already associated to a python object of type FakeTensor`

Reply:
In the end we might want to instead fix the callsite :)

Reply:
For context, we've found below as one of the callsite. The empty tensor was created as a fake tensor under user created fake mode context. It can be repro-ed shortly with
```python
with fake_mode:
    FakeTensor(fake_mode, torch.empty(...))
```

https://github.com/pytorch/pytorch/blob/14d0b6e0d8d2a087ff06ac7e5a1be5f4a64a7d0f/torch/_subclasses/fake_tensor.py#L779-L788

Reply:
would replace `self = torch.Tensor._make_subclass()` by `fake_mode.from_tensor()` an option?

Reply:
from fake tensor design doc:

> Fake tensor of fake tensor?
There is interest in sending fake tensors as user inputs into the PT2 stack, which would imply we would need to be able to create a fake tensor of a fake tensor. This isn’t really supported right now, but maybe it would not be too difficult to do.

I have no idea how to get around that in a different way. Ideas?

Reply:
Didn't we agree that you were going to do the "make PT2 reuse the ambient fake tensor context" strategy

Reply:
@ezyang  We did! In fact, `torch/_dynamo/eval_frame.py::export` now takes a [fake_mode from user](https://github.com/pytorch/pytorch/pull/100017/files#diff-c663df72d7a648204ff256e604fe3c9b393e7e2d58204e3bc818d90a99712911R739) which is ultimately consumed in two places:

The first place is inside [OutputGraph.\_\_init\_\_](https://github.com/pytorch/pytorch/pull/100017/files#diff-1bbe80a42da83f0b0e1386c3aca92323930626a65173c7d7e838931446fb4324R229 and the second place is within [dynamo_normalization_capturing_compiler](https://github.com/pytorch/pytorch/pull/100017/files#diff-c663df72d7a648204ff256e604fe3c9b393e7e2d58204e3bc818d90a99712911R851)

Should this solve the issue mentioned here? Seems unrelated, because the user data will already be fakeifed and when the input is fakeifed again, internally, `FakeTensor.\_\_new\_\_` will be called and the issue `RuntimeError: Creating a new Tensor subclass FakeTensor but the raw Tensor object is already associated to a python object of type FakeTensor` will still happen

Reply:
For reference, the issue with `make_subclass`comes from [these checks](https://github.com/pytorch/pytorch/blob/main/torch/csrc/autograd/python_variable.cpp#L1791-L1864) at `THPVariable_NewWithVar` which is called by `THPVariable_make_subclass`

Reply:
Do you have a small repro you could post of what the failing pattern is ?

Reply:
So I think what is probably happening is something like, we are calling "detach()" to generate a fresh tensor object, but somehow the FakeTensor dispatch mode is on and interposing on the detach() call which fakeifies the object. We then are unable to create the new tensor without overwriting the old subclass.

Reply:
> Do you have a small repro you could post of what the failing pattern is ?

Yes, I do

```python
import torch

fake_mode = torch._subclasses.FakeTensorMode()
with fake_mode:
    fake_x = torch.randn(1,2,3)
```

which leads to

```bash
Traceback (most recent call last):
  File "/home/thiagofc/dev/pytorch_repros/fake_from_fake.py", line 7, in <module>
    fake_from_fake_x = torch._subclasses.FakeTensor(fake_mode, fake_x, fake_x.device)
  File "/home/thiagofc/dev/github/pytorch-dev1/torch/_subclasses/fake_tensor.py", line 904, in __new__
    self = torch.Tensor._make_subclass(
RuntimeError: Creating a new Tensor subclass FakeTensor but the raw Tensor object is already associated to a python object of type FakeTensor
```

Reply:
That doesn't error on master, it only errors with your changes. Do you have a small repro of what you are trying to that errors on master?

Reply:
Weird, I just tried with latest nightly and it repro'ed. It also reproed on CI before changing fake_tensor.py to create a fake tensor from another. here is some extra info

```bash
(dev2-cu113-py39) ✘-1 ~/dev/github/pytorch-dev2 [main ↑·1|…3⚑ 1] 
13:59 $ conda list | grep -i torch
magma-cuda113             2.5.2                         1    pytorch
pytorch-triton            2.1.0+7d1a95b046          pypi_0    pypi
torch                     2.1.0.dev20230518+cu118          pypi_0    pypi
torchaudio                2.1.0.dev20230518+cu118          pypi_0    pypi
torchvision               0.14.0a0+3c9ae0a           dev_0    <develop>
(dev2-cu113-py39) ✔ ~/dev/github/pytorch-dev2 [main ↑·1|…3⚑ 1] 
13:59 $ python ~/dev/pytorch_repros/fake_from_fake.py 
Traceback (most recent call last):
  File "/home/thiagofc/dev/pytorch_repros/fake_from_fake.py", line 7, in <module>
    fake_from_fake_x = torch._subclasses.FakeTensor(fake_mode, fake_x, fake_x.device)
  File "/home/thiagofc/miniconda3/envs/dev2-cu113-py39/lib/python3.9/site-packages/torch/_subclasses/fake_tensor.py", line 904, in __new__
    self = torch.Tensor._make_subclass(
RuntimeError: Creating a new Tensor subclass FakeTensor but the raw Tensor object is already associated to a python object of type FakeTensor
(dev2-cu113-py39) ✘-1 ~/dev/github/pytorch-dev2 [main ↑·1|…3⚑ 1] 
13:59 $ cat ~/dev/pytorch_repros/fake_from_fake.py 
import torch

fake_mode = torch._subclasses.FakeTensorMode()
with fake_mode:
    fake_x = torch.randn(1,2,3)

fake_from_fake_x = torch._subclasses.FakeTensor(fake_mode, fake_x, fake_x.device)(dev2-cu113-py39) ✔ ~/dev/github/pytorch-dev2 [main ↑·1|…3⚑ 1] 
13:59 $ 
```

Reply:
`fake_from_fake_x` wasn't in the repro. now i've got it!

Reply:
when would you want to use `fake_from_fake_x` instead of the existing `fake_x` ?

Reply:
We want to develop a brand new ONNX exporter (to replace the current TorchScript-based) which does symbolic tracing on model+input. This is important to export huge models

An example usage would be:

```python
# Fakefying input/model before calling torch._dynamo.export
with fake_mode:
    fake_x = torch.rand(5, 2, 2)
    fake_model = Model()

# Calling torch._dynamo.export without active fake mode
graph_module, guards = torch._dynamo.export(
    fake_model,
    fake_x,
    aten_graph=True,
    fake_mode=fake_mode
)
```

The full context is on the PR description, but let me know if anything is missing or confusing

Reply:
@thiagocrepaldi but we shouldn't do this, right? If we're preserving the fake mode, we also want to preserve the input fake tensors, so we shouldn't end up fakeifying from an already fake tensor.

Actually this might be a bit troublesome, because it means if we mutate metadata on the fake tensors, that mutation will get reflected outside of the export. But we can figure that out elsewise.

Reply:
> Actually this might be a bit troublesome, because it means if we mutate metadata on the fake tensors, that mutation will get reflected outside of the export

IMO we should still be converting input FakeTensors to FakeTensors of a new mode. There are a couple reasons to do this :

1. Mutation as you have mentioned 
2. In dynamo tracing we trace with `allow_non_fake_inputs=False`. There have been many instances where we did fully unwrap a python call (see below) where our failure to convert inputs to fake tensors also revealed that we did not properly inline some piece of python logic and therefore add appropriate guards etc. I think we should have the same guardrails for fake tensor inputs. 


Example errors: https://github.com/pytorch/pytorch/issues/96411 and https://github.com/pytorch/pytorch/issues/96362

Reply:
> IMO we should still be converting input FakeTensors to FakeTensors of a new mode.

I agree this is better but this is a biiiig refactor. Dynamo in particular makes hella assumptions everywhere.

Also, there are use cases for reusing the fake mode as is; for example, this is what @jansel wanted for the fx tracing for subgraph matching.